### PR TITLE
Fix game stats division by zero

### LINKED ISSUE #559: Game stats division by zero
Not entirely sure how `acus_per_player` is set to 0 (maybe there was an observer in the game?), but the `GameStatsService` doesn't handle this correctly.

```
Traceback (most recent call last): 
  File "/code/server/games/game.py", line 355, in _process_army_stats_for_player
    player, self, self._army_stats
  File "/code/server/stats/game_stats_service.py", line 79, in process_game_stats
    self._killed_acus(unit_stats, survived, a_queue)
  File "/code/server/stats/game_stats_service.py", line 203, in _killed_acus
    self._increment(ACH_DONT_MESS_WITH_ME, int(killed_acus / acus_per_player), achievements_queue)
ZeroDivisionError: division by zero
```

### PR DESCRIPTION
**Pull request by devopsdeluxe**
Fixes #559

### CODE REVIEW

**Comment by devopsdeluxe** [comment]
I would like to create a test case for this issue before it is merged.

**Comment by Askaholic** [comment]
Good idea. There is also a "draft" feature for PR's that will make sure the PR can't get merged before its ready, but I think you have to select it when you create the PR

**Comment by cleborys** [comment]
In the top right under "Reviewers" there is a little inconspicuous "convert to draft" link to click.

**Comment by cleborys** [comment]
Do you know *why* there is a division by `acus_per_player` in the first place?
My only guess was that you could have less than one acu per player with the shared armies mod, but even then I don't understand why it's not just incremented by `killed_acus`.

**Comment by devopsdeluxe** [comment]
> Do you know why there is a division by acus_per_player in the first place?

No.

>  but even then I don't understand why it's not just incremented by killed_acus.

I'm in agreeance, but I don't think I have the authority to make that decision. ;)

**Comment by Askaholic** [comment]
> I'm in agreeance, but I don't think I have the authority to make that decision. :P

The achievements stuff has been broken for so long, you should feel free to propose a fix if you find stuff like this.

**Comment by devopsdeluxe** [comment]
@Askaholic what's the proper outlet to suggest a change (GitHub issue, Zulip, Discord)?

> The achievements stuff has been broken for so long, you should feel free to propose a fix if you find stuff like this.

Also, TIL. I'll keep that in mind going forward!

**Comment by Askaholic** [comment]
Zulip is a good place, or you could make a commit and ask for input on github. For this one, I'd say just go ahead and commit the change so we can review it, since it sounds like you and @cleborys both came to the same conclusion.

File: server/stats/game_stats_service.py [hunk 1]
```diff
@@ -193,15 +193,12 @@ def _faction_played(self, faction, survived, achievements_queue, events_queue):
...
-        # I'm aware that this is not perfectly correct, but it's edge case and tracking who got killed by whom would
-        # require game code changes
-        if killed_acus >= 3 * acus_per_player and survived:
-            self._unlock(ACH_HATTRICK, achievements_queue)
-
-        self._increment(ACH_DONT_MESS_WITH_ME, int(killed_acus / acus_per_player), achievements_queue)
+        if killed_acus > 0:
+            self._increment(ACH_DONT_MESS_WITH_ME, killed_acus, achievements_queue)
+            if killed_acus >= 3 and survived:
+                self._unlock(ACH_HATTRICK, achievements_queue)
```
Comment (cleborys) [inline server/stats/game_stats_service.py:201]: I think these two lines could be unindented one step

Comment (devopsdeluxe) [reply server/stats/game_stats_service.py:201]: My reasoning behind nesting them there was that if `killed_acus` is not greater than 0, the second `if` statement would not to be evaluated.

If you would still like them unindented one step, let me know and I'll update the MR.

Comment (cleborys) [reply server/stats/game_stats_service.py:201]: Sure, your choice :)

Comment (devopsdeluxe) [reply server/stats/game_stats_service.py]: > I'd be surprised if there was much of a difference in performance here.

Certainly nothing noticeable. :P

**Comment by devopsdeluxe** [comment]
Sure thing!

File: tests/unit_tests/test_game_stats_service.py [hunk 2]
```diff
@@ -114,9 +114,7 @@ def unit_stats():
     }
 
 
-async def test_process_game_stats(
-    game_stats_service, event_service, achievement_service, player, game
-):
+async def test_process_game_stats(game_stats_service, event_service, achievement_service, player, game):
```
Comment (Askaholic) [inline tests/unit_tests/test_game_stats_service.py:117]: Keep lines shorter than 80 chars unless there is a good formatting reason.

Comment (devopsdeluxe) [reply tests/unit_tests/test_game_stats_service.py]: Wrapped the lines!

I will defend using an extra 4 spaces for the function arguments though, as it places them on a different indentation than the function's body.

Comment (Askaholic) [reply tests/unit_tests/test_game_stats_service.py]: The 8 space indentation has come up before and we decided against it. I recently added a [CONTRIBUTING.md](https://github.com/FAForever/server/blob/develop/CONTRIBUTING.md) file with some style examples.

You don't necessarily need to put the arguments all on their own line for functions which will rarely change like these. It does make merging a lot easier when there are conflicts and it gives more space for typing.

Comment (devopsdeluxe) [reply tests/unit_tests/test_game_stats_service.py]: I've reverted the 8 space tabs, and cleaned up a few other minor things I noticed.

Let me know if there's anything else you'd like changed.

File: tests/unit_tests/test_game_stats_service.py [hunk 3]
```diff
@@ -520,48 +524,62 @@ async def test_killed_acus_one_and_died(
...
 
+async def test_killed_acus_three_and_died(
+    game_stats_service, achievement_service, event_service, unit_stats
+):
+    unit_stats['cdr']['kills'] = 3
+    unit_stats['cdr']['lost'] = 1
+
+    game_stats_service._killed_acus(unit_stats, False, [])
+    achievement_service.increment.assert_called_once_with(
+        ach.ACH_DONT_MESS_WITH_ME, 3, [])
```
Comment (Askaholic) [inline tests/unit_tests/test_game_stats_service.py:535]: I personally like to align the closing brackets like this
```suggestion
        ach.ACH_DONT_MESS_WITH_ME, 3, []
    )
```

**Pull request merged**